### PR TITLE
Maddie Shields & Hayden Williams - Edges - oo - RideShare

### PR DESCRIPTION
**Pull request by madaleines**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | A design decision made was in the trip dispatcher and we assign the "first_available_driver" and how that was a method which called upon multiple helper methods to determine assigning a driver who has no trips, otherwise the driver who hadn't taken a trip recently, and lastly if none were available to Raise an error. The alternative would have been keeping the code in one block which would have made it hard to read.
Describe any examples of composition that you encountered in this project, if any | Instead of having all functionality in the TripDispatcher class, we can see it broken up in multiple classes with respect to Driver, User and Trip. When the TripDispatcher creates a new Trip it can call upon each class for the desired functionality.
Describe the relationship between `User` and `Driver`| User is the parent class to Driver. Driver inherits the qualities of a User as well needing a vehicle_id, driven_trips array, and status. A passenger could potentially be a driver as well and have the same id.
Describe a _nominal test_ that you wrote for this assignment. | All methods which calculated totals needed testing to make sure they returned a float and a specific amount.
Describe an _edge case test_ that you wrote for this assignment | A driver not being able to drive themselves
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | We got to a part of the request_trip method and I began getting confused about adding a trip to the driver's driven trips, the passenger trips, and the trips array which Hayden helped break up the relationships to me. Hayden received more insight on how to approach TDD.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We were both very direct and also able to recognize quickly that feedback received was not personal. This helped us push through fixing problems as they arose fairly quickly.

### CODE REVIEW

File: lib/trip.rb [hunk 1]
```diff
@@ -2,25 +2,30 @@
...
       @end_time = input[:end_time]
       @cost = input[:cost]
       @rating = input[:rating]
 
-      if @rating > 5 || @rating < 1
-        raise ArgumentError.new("Invalid rating #{@rating}")
-      end
+      raise ArgumentError.new("Invalid rating #{@rating}") if @rating != nil && (@rating > 5 || @rating < 1)
+
+      raise ArgumentError.new("Start time cannot be greater than End time") if @end_time != nil && (@end_time < @start_time)
```
Comment (droberts-sea) [inline lib/trip.rb:18]: For both these lines, I think the postfix conditional makes the code harder to read, not easier. A good rule of thumb is to keep lines shorter than 80 characters, but even that is pushing it.

File: specs/trip_spec.rb [hunk 2]
```diff
@@ -40,5 +45,31 @@
         }.must_raise ArgumentError
       end
     end
+
+    it "raises an error if end time precedes start time" do
+      start_time = Time.parse('2015-05-20T12:14:00+00:00')
+      end_time = start_time - 25 * 60 # 25 minutes
+      trip_data = {
```
Comment (droberts-sea) [inline specs/trip_spec.rb:52]: What does this method do for an in-progress trip (`nil` end time)?

File: specs/user_spec.rb [hunk 3]
```diff
@@ -59,4 +77,16 @@
       end
     end
   end
+
+  describe '#net_expenditures method' do
+    it 'calculates the total expenditures by a User' do
+      expect(@user.net_expenditures).must_equal 33.70
+    end
```
Comment (droberts-sea) [inline specs/user_spec.rb:84]: You're missing a couple of interesting cases for both these new `User` methods:

- What happens if the user has no trips?
- What happens if the user has an incomplete trip?

File: lib/user.rb [hunk 4]
```diff
@@ -16,5 +14,21 @@ def initialize(input)
     def add_trip(trip)
       @trips << trip
     end
+
+    def net_expenditures
+      total_expenditures = 0
+      @trips.each do |trip|
+        total_expenditures += trip.cost
```
Comment (droberts-sea) [inline lib/user.rb:21]: This doesn't handle incomplete trips (end time, cost and rating are `nil`)!

File: lib/driver.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
+require_relative 'user'
+
+STATUS_OPTIONS = [:AVAILABLE, :UNAVAILABLE]
+DRIVERS_CUT = 0.8
+FEE = 1.65
```
Comment (droberts-sea) [inline lib/driver.rb:5]: Great use of constants to hold these values!

File: lib/driver.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    def add_driven_trip(trip)
+      raise ArgumentError unless trip.is_a? Trip
+      @driven_trips << trip
+    end
+
+    def average_rating
+      trip_ratings = []
+      @driven_trips.each do |trip|
+        trip_ratings << trip.rating
```
Comment (droberts-sea) [inline lib/driver.rb:29]: All of these methods will hit an error if you try to run them on a driver with an incomplete trip. There are a couple of workarounds for this.

You could explicitly ignore trips with a `nil` cost:

```ruby
@driven_trips.each do |trip|
  if trip.cost.nil?
    next
  end
  # ... add to the total ...
end
```

Or even better, you could write a helper method that returns a list of complete trips:

```ruby
def completed_trips
  return @driven_trips.reject { |t| t.end_time.nil? }
end

def total_revenue
  completed_trips.each do |trip|
    # ... same logic as before ...
  end
end
```

File: specs/driver_spec.rb [hunk 7]
```diff
@@ -77,28 +143,70 @@
...
+    end
+
+    it 'calculates the total expenditures by a Driver' do
+      expect(@driver.net_expenditures).must_equal 21.00
+    end
+
+    it 'returns a negative number when total revenue is greater than gross expenditures' do
+      @driver.add_driven_trip(@trip_5)
+      expect(@driver.net_expenditures).must_equal -6.08
+    end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:193]: This is a great edge case. I would also be interested in:

- No driven trips
- No trips as a passenger
- No trips at all

File: specs/driver_spec.rb [hunk 8]
```diff
@@ -77,28 +143,70 @@
...
+      @status = @new_trip.driver.status
+    end
+
+    it "adds new in progress trip from TripDispatcher#request_trip to driver's collection of trips" do
+      expect(@driven_trips.include?(@new_trip)).must_equal true
+    end
+
+    it "sets driver's status to :UNAVAILABLE" do
+      expect(@status).must_equal :UNAVAILABLE
+    end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:210]: What if the driver was already unavailable?

File: specs/driver_spec.rb [hunk 9]
```diff
@@ -77,28 +143,70 @@
...
   describe "total_revenue" do
-    # You add tests for the total_revenue method
+    before do
+      @driver.add_driven_trip(@trip_1)
+      @driver.add_driven_trip(@trip_2)
+    end
+    it "returns 0 if the driver has made no trips" do
+      driver = RideShare::Driver.new(
+        id: 54,
+        name: "Rogers Bartell IV",
```
Comment (droberts-sea) [inline specs/driver_spec.rb:168]: I like that you got the case where there are no trips.

File: lib/trip_dispatcher.rb [hunk 10]
```diff
@@ -1,70 +1,160 @@
...
+
+    def check_drivers_not_passenger?(user_id)
+      all_available_drivers = generate_available_drivers
+
+      available_drivers = all_available_drivers.find_all do |driver|
+        driver.id != find_passenger(user_id).id
+      end
+
+      raise ArgumentError.new("NO DRIVERS AVAILABLE") if available_drivers.empty?
+
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:104]: None of the other helper methods for finding a driver raise an error - what makes this one special.

Also, I'm not sure I would use an `ArgumentError` here, because we didn't get here as the result of the user passing in a bad argument. Defining your own custom exception would be appropriate here. Maybe something like a `NoAvailableDriversError`.

File: lib/trip_dispatcher.rb [hunk 11]
```diff
@@ -1,70 +1,160 @@
...
+    end
+
     def find_passenger(id)
       check_id(id)
       return @passengers.find { |passenger| passenger.id == id }
     end
 
+    def generate_available_drivers
+      available_drivers = @drivers.find_all do |driver|
+        driver.status == :AVAILABLE
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:92]: I love the way you broke the complex logic out of finding a driver into many small helper methods - this makes the resulting code much easier to reason about.

File: lib/trip_dispatcher.rb [hunk 12]
```diff
@@ -1,70 +1,160 @@
...
+      end
+    end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:94]: I would like to see an explicit return here.

File: lib/trip_dispatcher.rb [hunk 13]
```diff
@@ -1,70 +1,160 @@
...
+      available_drivers = check_drivers_not_passenger?(user_id)
+      return available_drivers.find { |driver| driver.driven_trips.empty? }
+    end
+
+    def find_oldest_trip_driver(user_id)
+      available_drivers = check_drivers_not_passenger?(user_id)
+
+      find_most_recent_trips = available_drivers.map do |driver|
+        driver.driven_trips.max_by(&:end_time)
+      end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:118]: The logic on line 117 might work well as a helper method in `Driver`, something like `Driver.most_recent_driven_trip`

File: specs/trip_dispatch_spec.rb [hunk 14]
```diff
@@ -111,4 +95,67 @@
...
+  describe '#request_trip method' do
+    before do
+      @requested_trip = @dispatcher.request_trip(1)
+    end
+
+    it "generates an array of :AVAILABLE drivers" do
+      available_drivers = @dispatcher.generate_available_drivers
+      expect(available_drivers).must_be_kind_of Array
+      expect(available_drivers.first.status).must_equal :AVAILABLE
+      expect(available_drivers.last.status).must_equal :AVAILABLE
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:108]: I would go further than just checking the first and the last here: iterate through the list and check each one.

File: specs/trip_dispatch_spec.rb [hunk 15]
```diff
@@ -111,4 +95,67 @@
...
+    end
+
+    it "returns an array of :AVAILABLE drivers that aren't the Passenger" do
+      available_drivers = @dispatcher.check_drivers_not_passenger?(2)
+      expect(available_drivers).must_be_kind_of Array
+
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:114]: Is driver 2 usually available?

File: specs/trip_dispatch_spec.rb [hunk 16]
```diff
@@ -111,4 +95,67 @@
...
+      available_drivers.each do |driver|
+        expect(driver.id).wont_equal 2
+      end
+    end
+
+    it "Assigns the first :AVAILABLE driver with no driven trips to a new Trip" do
+      driver  = @requested_trip.driver
+      expect(driver).must_be_instance_of RideShare::Driver
+      expect(driver.driven_trips.length).must_equal 1
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:123]: None of these tests answer the question: what do these helper methods do if no suitable driver is available?

File: specs/trip_dispatch_spec.rb [hunk 17]
```diff
@@ -111,4 +95,67 @@
...
+      expect(@dispatcher.trips.last).must_equal @requested_trip
+    end
+
+    it "Does not assign a driver with same ID as passenger" do
+      new_trip = @dispatcher.request_trip(8)
+      driver_id = new_trip.driver.id
+      passenger_id = new_trip.passenger.id
+
+      expect(driver_id).wont_equal passenger_id
+      expect(driver_id).must_equal 5
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:148]: Would it normally assign driver 8 in this case? How do you know? How could you make it clear to the reader of this test?

A comment explicitly calling out this assumption would be a good start, but even better would be some assertions before your code that verify the expected behavior. For example, you might call `@dispatcher.first_available_driver` with a different passenger ID, and make sure that you get driver 8 back.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes
**Wave 1** |
Appropriate use of Ruby's `Date`| yes
`Trip` has a helper method to calculate duration| yes
`User` (passenger) has a method to calculate total cost of all trips| yes
Tests for wave 1 | missing some edge cases - see inline
**Wave 2** |
`Driver` inherits from `User` | yes
`Driver` has `add_driven_trip` method | yes
`Driver` has method to calculate average rating | yes
`Driver` has method to calculate net expenditures and it uses `super` | yes
`Driver` has a method to calculate total revenue | yes
Tests for wave 2 | see inline
**Wave 3** |
`TripDispatcher` has a new method to create trips | yes
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | yes
Complex logic was correctly implemented | yes
Tests for `request_trip` | yes - see inline
Methods from wave 1 and 2 handle incomplete trips | no
Tests for wave 1 and 2 methods with incomplete trips | no
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | yes
Appropriate helper methods were made to help with complex logic | yes
Tests for wave 4 | yes - see inline
**Overall** |

Good job overall. Your product code is strong, and it's clear you have a good understanding of the problem. However, I see room for improvement around testing. This manifests in two big ways:
- Missing edge cases (user with no trips, or with an incomplete trip)
- Unclear and unverified assumptions (driver 8 would normally be assigned to this trip)

However, this is the first time you've had to write a substantial portion of the tests, so I'm particularly concerned (especially given how last week went in general). I've tried to call out missing tests and places where things could be improved inline below. Please keep these in mind as you work on the hotel project this week, and keep up the hard work!
